pkg/apis/batch/types.go

```diff
@@ -251,6 +252,30 @@ type JobStatus struct {
 	// represented as "1,3-5,7".
 	// +optional
 	CompletedIndexes string
+
+	// UncountedPodUIDs holds the UIDs of Pods that have terminated but the job
+	// controller hasn't yet accounted for in the status counters.
+	// If nil, it means that the job controller is not using this field for
+	// tracking job completion.
```

Comment:
There's a mismatch between this description and what the storage layer does. This says when it's nil the controller doesn't use this feature, which means when the feature was on and job was created and then the feature was off the field should be nil again, right? So you either:
1. change the description to account for the above situation
2. change the storage layer (`pkg/registry/batch/job/strategy.go`) such that in `PrepareForUpdate` it will set the field to nil.

Reply:
The opposite of 2nd scenario could also happen, for example `oldObject.field != nil and featureGate is not enabled` momentarily in the old apiserver which has not been rolledout, we would preserve the value associated with oldObj and controller may act upon. To me, kube is eventually consistent and there will be stages which may not be desired in reconciliation. The entities waiting on the job workloads should wait for certain time before those reconciliation happens or timeout.

 If you think null is a condition we explicitly in other scenarios, then perhaps it is ok to put it in the comments, if not I'd drop it.

Reply:
I like that suggestion from @ravisantoshgudimetla and I think we are using that in other places already.

Reply:
Actually, on a second thought - yeah I think @alculquicondor point is strong enough, but we need to make sure this is properly documented :sweat_smile:

Reply:
Anything you would like to be added to the comment? I still think the description is accurate of what happens.

Reply:
After having a private conversation with @alculquicondor this LGTM. However the contention point of `overloaded meaning of empty struct in the status` is still valid.

@alculquicondor told that he will think of alternatives during beta stage.

Reply:
The idea is that, if there is a way to start tracking with finalizers an existing job that was previously tracked without them, there is no need for distinguishing nil from empty struct.

If we cannot find a safe way to do that, the distinction can still be removed once we figured that the tracking has existed for long enough that there shouldn't exist jobs tracked with the legacy algorithm.

Reply:
@ravisantoshgudimetla proposed one solution to avoid having to distinguish between empty and nil: add the finalizer to the Job through an admission controller.

The updated proposal is in kubernetes/enhancements#2776